### ISSUE TITLE
allocator: thrashing behavior when replication factor is set to 1

### ISSUE THREAD

**Issue by wenyihu6** [cockroachdb/cockroach#108420]
**Describe the problem**

When replication is set to one and the removal target is the leaseholder, the allocator rebalances by adding a new replica but we are not removing. And we are waiting for the range to be touched next time to perform lease transfer and remove the old replica. This leads to thrashing behavior and over-replication sometimes.

https://github.com/cockroachdb/cockroach/blob/9ed70271734d0871254626d7909cb48fcb348d78/pkg/kv/kvserver/allocator/plan/util.go#L54-L62

**To Reproduce**

What did you do? Describe in your own words.

https://github.com/cockroachdb/cockroach/blob/5e1c94962945edf9d1ce134c0f8e6e4340a5c896/pkg/kv/kvserver/asim/tests/testdata/rand/rand_ranges#L238C1-L272C1

Jira issue: CRDB-30459

**Comment by kvoli** [comment]
To add to this issue:

**Describe the problem**

Leases may thrash back and forth between nodes when the replication factor is 1 and there are multiple nodes (presumably worse when there are just 2):

![image](https://github.com/cockroachdb/cockroach/assets/39606633/7c2bf2e0-241f-44cc-be4f-8485f72cf2a2)
![image](https://github.com/cockroachdb/cockroach/assets/39606633/64f3ebeb-cd19-4366-a907-4b746c96a702)
![image](https://github.com/cockroachdb/cockroach/assets/39606633/3f629356-caad-4a89-aa42-911f43d92219)


Or simulated:


<details><summary>Details</summary>
<p>

20 minute simulated run with 100 ranges RF=1 and 2 nodes:

```
100.00 ┼──────╮      ╭───╮
 93.33 ┤      ╰╮    ╭╯   ╰────╮
 86.67 ┤       │    │         │
 80.00 ┤       ╰╮  ╭╯         ╰╮    ╭─────╮
 73.33 ┤        │  │           │   ╭╯     ╰─╮   ╭───╮
 66.67 ┤        ╰╮╭╯           ╰╮ ╭╯        ╰╮ ╭╯   ╰─╮  ╭──╮
 60.00 ┤         ││             │ │          │ │      ╰╮╭╯  ╰──╮ ╭──╮
 53.33 ┤         ╭╯             ╰╮╯          ╰╭╯       │╯      ╰╭╯  ╰──╮───────╭───────
 46.67 ┤         │╮             ╭╰╮          ╭╯╮       ╰╮      ╭╯╮  ╭──╰───────╯───────
 40.00 ┤         ││             │ │          │ │      ╭╯╰╮  ╭──╯ ╰──╯
 33.33 ┤        ╭╯╰╮           ╭╯ ╰╮        ╭╯ ╰╮   ╭─╯  ╰──╯
 26.67 ┤        │  │           │   ╰╮     ╭─╯   ╰───╯
 20.00 ┤       ╭╯  ╰╮         ╭╯    ╰─────╯
 13.33 ┤       │    │         │
  6.67 ┤      ╭╯    ╰╮   ╭────╯
  0.00 ┼──────╯      ╰───╯
                                             lease
```

</p>
</details> 


**To Reproduce**

```
roachprod create cluster local -n 2
roachprod start cluster local

ALTER RANGE default CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER DATABASE system CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER RANGE meta CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER RANGE system CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER RANGE liveness CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.replication_constraint_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.replication_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.statement_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.transaction_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.tenant_usage CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.span_stats_tenant_boundaries CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.statement_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';
ALTER TABLE system.public.transaction_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';

ALTER RANGE default CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER DATABASE system CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER RANGE meta CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER RANGE system CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER RANGE liveness CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.replication_constraint_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.replication_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.statement_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.transaction_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.tenant_usage CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.span_stats_tenant_boundaries CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.statement_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
ALTER TABLE system.public.transaction_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[]';
```


**Expected behavior**

Leases do not thrash for an extended duration (>1 minute). 



**Additional data / screenshots**

This is caused by hitting the rebalance code path which first adds a replica when RF=1:

https://github.com/cockroachdb/cockroach/blob/61efb1e71d2d8b0a20e971afca8b2f002cd7cc76/pkg/kv/kvserver/allocator/plan/util.go#L39-L39

Which then requires a lease transfer to occur from voter removal:

https://github.com/cockroachdb/cockroach/blob/61efb1e71d2d8b0a20e971afca8b2f002cd7cc76/pkg/kv/kvserver/allocator/plan/replicate.go#L646-L647

However, there isn't a base case where the lease will stop transferring reliably -- which can cause lease thrashing seen above. Ideally, we'd avoid transferring the lease back to a replica if we recently received it from them, and there isn't a hard requirement

**Comment by kvoli** [comment]
This appears fixable by [lifting](https://github.com/cockroachdb/cockroach/issues/124171#issuecomment-2115351874) the restriction that 1 voter replica rebalance operations are converted into an addition:


<details><summary>repro</summary>
<p>

```
roachprod destroy local
roachprod create local -n 2
roachprod start local

roachprod sql local:1 -- -e "CREATE TABLE t (i INT PRIMARY KEY);"
roachprod sql local:1 -- -e "INSERT INTO t select generate_series(1,1000);"

roachprod sql local:1 -- -e "ALTER TABLE t CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER RANGE default CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER DATABASE system CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER RANGE meta CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER RANGE system CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER RANGE liveness CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.replication_constraint_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.replication_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.statement_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.transaction_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.tenant_usage CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.span_stats_tenant_boundaries CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.statement_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.transaction_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[+node1]';"
sleep 10

roachprod sql local:1 -- -e "ALTER TABLE t SPLIT AT SELECT i FROM t;"

roachprod sql local:1 -- -e "ALTER TABLE t CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER RANGE default CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER DATABASE system CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER RANGE meta CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER RANGE system CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER RANGE liveness CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.replication_constraint_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.replication_stats CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.statement_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.transaction_statistics CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.tenant_usage CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.span_stats_tenant_boundaries CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.statement_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
roachprod sql local:1 -- -e "ALTER TABLE system.public.transaction_activity CONFIGURE ZONE USING num_replicas=1, constraints = '[]';"
```

</p>
</details> 

![image](https://github.com/cockroachdb/cockroach/assets/39606633/d0e864c8-1a1e-4d33-b933-f9c923d3914f)
![image](https://github.com/cockroachdb/cockroach/assets/39606633/ff3c6515-5211-494a-ac3b-0442aa2a29c5)

**Comment by kvoli** [comment]
Will be closed on backport to 23.1, 23.2 and 24.1.

**Comment by kvoli** [comment]
Backports:
- #124453
- #124487
- #124500
